### DISCUSSION TITLE
Client Credentials Flow openAPI UI

### DISCUSSION THREAD

**Discussion by MarlieChiller** [fastapi/fastapi#7846] [Questions] [answered]
### First check

* [x] I used the GitHub search to find a similar issue and didn't find it.
* [x] I searched the FastAPI documentation, with the integrated search.
* [x] I already searched in Google "How to X in FastAPI" and didn't find any information.

### Description

As far as my searches could tell, I could not find any solid information on how to go about doing this (i may be missing something obvious):

I currently have an api using the documented password oauth flow, however I am now in the process of upgrading this flow to something more robust as the password flow is not recommended in most cases. I have selected a client_credentials type flow and have built an authorization server to generate JWT tokens. Now, in order to 'log in' via the openAPI UI, one must retrieve a JWT token from this authorization server which can be decoded to access relevant endpoints. How can i modify the openAPI UI to utilise the usual 'authorize' button but instead of taking username & password, only take client_id, client_secret and scope. Is there a base class i can take that such as [this Oauth2 class](https://github.com/tiangolo/fastapi/blob/master/fastapi/security/oauth2.py#L114) which can be modified or used to hit the authorization server and retrieve a token?

Thanks for any help in advance

**Comment by MarlieChiller** [comment, marked as answer]
I think I found the solution for others looking to implement the code - tiangolo has already enabled this. instead of using:

```python
 oauth2_scheme = Oauth2PasswordBearer(
    tokenUrl=JWT_TOKEN_URL, scopes={"foo": "bar"}
)
```

you create your own version of the `Oauth2PasswordBearer` class like so:
```python
class Oauth2ClientCredentials(OAuth2):
    def __init__(
        self,
        tokenUrl: str,
        scheme_name: str = None,
        scopes: dict = None,
        auto_error: bool = True,
    ):
        if not scopes:
            scopes = {}
        flows = OAuthFlowsModel(clientCredentials={"tokenUrl": tokenUrl, "scopes": scopes})
        super().__init__(flows=flows, scheme_name=scheme_name, auto_error=auto_error)

    async def __call__(self, request: Request) -> Optional[str]:
        authorization: str = request.headers.get("Authorization")
        scheme, param = get_authorization_scheme_param(authorization)
        if not authorization or scheme.lower() != "bearer":
            if self.auto_error:
                raise HTTPException(
                    status_code=HTTP_401_UNAUTHORIZED,
                    detail="Not authenticated",
                    headers={"WWW-Authenticate": "Bearer"},
                )
            else:
                return None
        return param
```

with the key difference being a line change here:

`flows = OAuthFlowsModel(clientCredentials={"tokenUrl": tokenUrl, "scopes": scopes})`

to take the fastAPI model clientCredentials instead of Password. Hope that helps anyone else that gets stuck with this. It is actually in the documentation but it was not immediately apparent to me that I needed to do this. +1 for easy to understand source code :)

**Comment by tiangolo** [comment]
Thanks for reporting back and closing the issue @Charlie-iProov :+1: 

Yeah, there's some extra docs and even classes missing :sweat_smile: :memo:

**Comment by and-semakin** [comment]
Hey, @tiangolo, would you like to have a PR with this class? I'm ready to do it.

**Comment by matletix** [comment]
@Charlie-iProov Which class did you use for the form in your token endpoint ? I modified the `OAuth2PasswordRequestForm` like this :
```python
class OAuth2ClientCredentialsRequestForm:
    def __init__(
        self,
        grant_type: str = Form(None, regex="client_credentials"),
        scope: str = Form(""),
        client_id: str = Form(...),
        client_secret: str = Form(...),
    ):
        self.grant_type = grant_type
        self.scopes = scope.split()
        self.client_id = client_id
        self.client_secret = client_secret
```

But I get `422: Validation Error` errors when using the "Authorize" feature of the /docs interface. That's because the `client_id` and `client_secret` are put inside the Authorization header and not passed as requests parameters.

**Comment by jwhitlock** [comment]
You're almost there. Here's what [mine](https://github.com/mozilla-it/ctms-api/blob/main/ctms/auth.py) looks like:

```python
OAuth2ClientCredentialsRequestForm:
    def __init__(
        self,
        grant_type: str = Form(None, regex="client_credentials"),
        scope: str = Form(""),
        client_id: Optional[str] = Form(None),
        client_secret: Optional[str] = Form(None),
    ):
        self.grant_type = grant_type
        self.scopes = scope.split()
        self.client_id = client_id
        self.client_secret = client_secret
```

And then, in the [app view](https://github.com/mozilla-it/ctms-api/blob/main/ctms/app.py), handle credentials coming in the form body or the authorization header (via an optional ``HTTPBasicCredentials``):

```python
@app.post(
    "/token",
    include_in_schema=False,
    summary="Get OAuth2 access token",
    response_model=TokenResponse,
    responses={400: {"model": BadRequestResponse}},
)
def login(
    db: Session = Depends(get_db),
    form_data: OAuth2ClientCredentialsRequestForm = Depends(),
    basic_credentials: Optional[HTTPBasicCredentials] = Depends(token_scheme),
    token_settings=Depends(_token_settings),
):
    failed_auth = HTTPException(
        status_code=400, detail="Incorrect username or password"
    )

    if form_data.client_id and form_data.client_secret:
        client_id = form_data.client_id
        client_secret = form_data.client_secret
    elif basic_credentials:
        client_id = basic_credentials.username
        client_secret = basic_credentials.password
    else:
        raise failed_auth

    api_client = get_api_client_by_id(db, client_id)
    if not api_client or not api_client.enabled:
        raise failed_auth
    if not verify_password(client_secret, api_client.hashed_secret):
        raise failed_auth

    access_token = create_access_token(
        data={"sub": f"api_client:{client_id}"}, **token_settings
    )
    return {
        "access_token": access_token,
        "token_type": "bearer",
        "expires_in": int(token_settings["expires_delta"].total_seconds()),
    }
```

There may be a better way to implement this, by looking directly at the ``Request`` and extract credentials from either the form or the header, and pass it as a ``Depends`` rather than handling it in the view.

**Comment by matletix** [comment]
@jwhitlock Thank you, this works ! However it creates two authentication schemes in the `/docs` interface :

![image](https://user-images.githubusercontent.com/19894690/112117423-a5b24100-8bbb-11eb-931c-9b05b7b56b05.png)

So I used the `Request` object directly to access the Basic Auth credentials, something like :
```python
async def login(request: Request,
                form_data: OAuth2ClientCredentialsRequestForm = Depends()
):
    failed_auth = HTTPException(
        status_code=400, detail="Incorrect username or password"
    )

    basic_credentials = await HTTPBasic(auto_error=False)(request)

    if form_data.client_id and form_data.client_secret:
        client_id = form_data.client_id
        client_secret = form_data.client_secret
    elif basic_credentials:
        client_id = basic_credentials.username
        client_secret = basic_credentials.password
    else:
        raise failed_auth

    api_client = get_api_client_by_id(db, client_id)
    if not api_client or not api_client.enabled:
        raise failed_auth
    if not verify_password(client_secret, api_client.hashed_secret):
        raise failed_auth

    access_token = create_access_token(
        data={"sub": f"api_client:{client_id}"}, **token_settings
    )
    return {
        "access_token": access_token,
        "token_type": "bearer",
        "expires_in": int(token_settings["expires_delta"].total_seconds()),
    }
```

**Comment by MarlieChiller** [comment]
its been a while since i deep dived into this but i think you need to register your new class as a `Security()` argument like so: 

```python
oauth2_scheme = Oauth2ClientCredentials(
    tokenUrl=JWT_TOKEN_URL,
    scopes={
        "write": "write permission",
        "read": "read permission",
    },
)


async def validate_jwt(
    security_scopes: SecurityScopes, token: str = Security(oauth2_scheme)
):
```
where `validate_jwt` can be replaced with your `login` func. My knowledge is very hazy on this now though!

**Comment by jwhitlock** [comment]
I was able to avoid advertising ``HTTPBasic`` by removing my token endpoint from the API with this line:

```python
include_in_schema=False,
```

This feels like a hack, and makes me think looking at the ``Request`` object directly may make for a cleaner interface, similar to OAuth2PasswordBearer, but tuned for token requests:

https://github.com/tiangolo/fastapi/blob/5614b94ccc9f72f1de2f63aae63f5fe90b86c8b5/fastapi/security/oauth2.py#L152-L164

**Comment by MarlieChiller** [comment]
@jwhitlock that looks similar to what i have:

```python
async def validate_jwt(
    security_scopes: SecurityScopes, token: str = Security(oauth2_scheme)
):
    if security_scopes.scopes:
        authenticate_value = f'Bearer scope="{security_scopes.scope_str}"'
    else:
        authenticate_value = f"Bearer"
    try:
        payload = jwt.decode(
            token, JWT_KEY, algorithms=[JWT_ALGO], options={"verify_exp": True}
        )
        username: str = payload.get("sub")
        if username is None:
            raise http_401_exc(
                "invalid oauth2 credentials",
                authenticate_value,
                f"invalid oauth2 credentials for token: {token}",
            )
    except ExpiredSignatureError:
        raise http_401_exc(
            "JWT expired", authenticate_value, f"jwt token has expired: {token}"
        )
    except (PyJWTError, ValidationError):
        raise http_401_exc(
            "invalid JWT", authenticate_value, f"invalid jwt token: {token}"
        )

    for scope in security_scopes.scopes:
        if scope not in payload.get("scope", []):
            raise http_401_exc(
                "invalid scope provided",
                authenticate_value,
                f'{scope} not in {payload.get("scope", [])}',
            )
    return payload


def http_401_exc(detail: str, authentication_value: str, log_msg: str):
    log.error(log_msg)
    return HTTPException(
        status_code=status.HTTP_401_UNAUTHORIZED,
        detail=detail,
        headers={"WWW-Authenticate": authentication_value},
    )

**Comment by jwhitlock** [comment]
Sorry my comment about ``include_in_schema=False`` was directed at @matletix, who was seeing two authentication schemes in the API. I didn't read his response closely, I see that he found a way around it as well.

**Comment by lwangapatrick** [comment]
Hi I am trying to implement the same thing using client_id and client_secret. I am stuck on how to amend the OAuth2PasswordRequestForm class.  So far this is what I have.

`class Oauth2ClientCredentials(OAuth2):
    def __init__(
            self,
            tokenUrl: str,
            scheme_name: str = None,
            scopes: dict = None,
            auto_error: bool = True,
    ):
        if not scopes:
            scopes = {}
        flows = OAuthFlowsModel(clientCredentials={"tokenUrl": tokenUrl, "scopes": scopes})
        super().__init__(flows=flows, scheme_name=scheme_name, auto_error=auto_error)

    async def __call__(self, request: Request) -> Optional[str]:
        authorization: str = request.headers.get("Authorization")
        scheme, param = get_authorization_scheme_param(authorization)
        if not authorization or scheme.lower() != "bearer":
            if self.auto_error:
                raise HTTPException(
                    status_code=HTTP_401_UNAUTHORIZED,
                    detail="Not authenticated",
                    headers={"WWW-Authenticate": "Bearer"},
                )
            else:
                return None
        return param

OAuth2ClientCredentialsRequestForm:
def __init__(
        self,
        grant_type: str = Form(None, regex="client_credentials"),
        scope: str = Form(""),
        client_id: Optional[str] = Form(None),
        client_secret: Optional[str] = Form(None),
):
    self.grant_type = grant_type
    self.scopes = scope.split()
    self.client_id = client_id
    self.client_secret = client_secret

@app.post(
    "/token",
    include_in_schema=False,
    summary="Get OAuth2 access token",
    response_model=TokenResponse,
    responses={400: {"model": BadRequestResponse}},
)
def login(
        db: Session = Depends(get_db),
        form_data: OAuth2ClientCredentialsRequestForm = Depends(),
        basic_credentials: Optional[HTTPBasicCredentials] = Depends(token_scheme),
        token_settings=Depends(_token_settings),
):
    failed_auth = HTTPException(
        status_code=400, detail="Incorrect username or password"
    )

    if form_data.client_id and form_data.client_secret:
        client_id = form_data.client_id
        client_secret = form_data.client_secret
    elif basic_credentials:
        client_id = basic_credentials.username
        client_secret = basic_credentials.password
    else:
        raise failed_auth

    api_client = get_api_client_by_id(db, client_id)
    if not api_client or not api_client.enabled:
        raise failed_auth
    if not verify_password(client_secret, api_client.hashed_secret):
        raise failed_auth

    access_token = create_access_token(
        data={"sub": f"api_client:{client_id}"}, **token_settings
    )
    return {
        "access_token": access_token,
        "token_type": "bearer",
        "expires_in": int(token_settings["expires_delta"].total_seconds()),
    }


def verify_password(self, client_secret):
    return bcrypt.verify(client_secret, self.client_secret)
`

I have problems with the following lines;
What does this mean?
`api_client = get_api_client_by_id(db, client_id)`
where do i import OAuth2, OAuthFlowsModel, get_authorization_scheme_param  from as they are missing in my code?

**Comment by jwhitlock** [comment]
@lwangapatrick you may want to edit your comment, the code is not formatted well. See https://docs.github.com/en/github/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks for tips.

The missing imports:

```python
from fastapi.security.oauth2 import OAuth2, OAuthFlowsModel
from fastapi.security.utils import get_authorization_scheme_param
```

See https://github.com/tiangolo/fastapi/blob/master/fastapi/security/oauth2.py for the code that these code snippets are modeling.

**Comment by lwangapatrick** [comment]
As soon I sent the message. I figured out where the methods are located. But thank you anyways for your help.

**Comment by siva-ram** [comment]
where are TokenResponse, token_scheme and _token_settings defined in the above code?

**Comment by jwhitlock** [comment]
@siva-ram those are details of our implementation, and not needed. But, here they are.

First, ``TokenResponse``:

```python
from pydantic import BaseModel

class TokenResponse(BaseModel):
    """An OAuth2 Token response."""

    access_token: str
    token_type: str
    expires_in: int
```

I believe the purpose was to improve the API generated documentation. However, it seems better to exclude the token-requesting endpoint from the API docs. See the discussion above for a few ways to do that.

`token_scheme` is:

```python
from fastapi.security import HTTPBasic
token_scheme = HTTPBasic(auto_error=False)
```

This is the same ideas as [Simple HTTP Basic Auth](https://fastapi.tiangolo.com/advanced/security/http-basic-auth), but that code calls it ``security`` instead of ``token_scheme``. I liked the more precise name, and wanted to allow something else to be called ``security``.

``_token_settings`` allows us to override some settings:

```python
def _token_settings(
    settings: config.Settings = Depends(get_settings),
) -> Dict[str, Union[str, timedelta]]:
    return {
        "expires_delta": settings.token_expiration,
        "secret_key": settings.secret_key,
    }
```

`get_settings()` is basically the same as the one described in [Settings in a dependency](https://fastapi.tiangolo.com/advanced/settings/#settings-in-a-dependency). So, when run as an app, it gets these from the environment or defaults, and I can override just these two settings for tests.

Our defaults:

```python
from pydantic import BaseSettings

class Settings(BaseSettings):
    secret_key: str
    token_expiration: timedelta = timedelta(minutes=60)
    ...
```

This means that ``SECRET_KEY`` must be set or config will fail. We use a long random string for each deployment. This concept was borrowed from Django, see [SECRET_KEY](https://docs.djangoproject.com/en/4.0/ref/settings/#secret-key) for how it is used there.

**Comment by Riuzaky77** [comment]
Hey Guys! I have it all working Ok with the swagger docs with this. Hope can help you

```ruby
class Oauth2ClientCredentials(OAuth2):
    def __init__(
        self,
        tokenUrl: str,
        scheme_name: str = "oAuth2ClientCredentials",
        scopes: dict = None,
        auto_error: bool = True,
    ):
        if not scopes:
            scopes = {}
        flows = OAuthFlowsModel(
            clientCredentials={"tokenUrl": tokenUrl, "scopes": scopes}
        )
        super().__init__(flows=flows, scheme_name=scheme_name, auto_error=auto_error)

    async def __call__(self, request: Request) -> Optional[str]:
        authorization: str = request.headers.get("Authorization")
        scheme, param = get_authorization_scheme_param(authorization)
        if not authorization or scheme.lower() != "bearer":
            if self.auto_error:
                raise HTTPException(
                    status_code=status.HTTP_401_UNAUTHORIZED,
                    detail="Not authenticated",
                    headers={"WWW-Authenticate": "Bearer"},
                )
            else:
                return None
        return param

api_oauth2 = Oauth2ClientCredentials(
    tokenUrl=f"{settings.API_V1_STR}/login/access-token",
    scopes={scope: scope.description for scope in Scope.api_scopes()},
)

class HTTPClientCredentials(BaseModel):
    client_id: str
    client_secret: str

class OAuth2ClientCredentialsRequestForm:
    def __init__(
        self,
        grant_type: str = Form(None, regex="client_credentials"),
        scope: str = Form(""),
        client_id: Optional[str] = Form(None),
        client_secret: Optional[str] = Form(None),
    ):
        self.grant_type = grant_type
        self.scopes = scope.split()
        self.client_id = client_id
        self.client_secret = client_secret

class HTTPBasicClientCredentials(HTTPBasic):
    async def __call__(  # type:ignore
        self, request: Request
    ) -> Optional[HTTPClientCredentials]:
        basic = await super().__call__(request)
        if not basic:
            return None
        return HTTPClientCredentials(
            client_id=basic.username, client_secret=basic.password
        )

token_scheme = HTTPBasicClientCredentials(
    auto_error=False, scheme_name="oAuth2ClientCredentials"
)


@router.post("/access-token")
def login(
    *,
    form: OAuth2ClientCredentialsRequestForm = Depends(),
    basic_credentials: Optional[HTTPClientCredentials] = Depends(token_scheme),
    auth_service: AuthService = Depends(),
) -> schemas.ApiToken:
   
    if form.client_id and form.client_secret:
        client_id = form.client_id
        client_secret = form.client_secret
    elif basic_credentials:
        client_id = basic_credentials.client_id
        client_secret = basic_credentials.client_secret
    else:
        HTTPException(status_code=400, detail="Client credentials not provided")
    pass

```

**Comment by bdelpey** [comment]
> @siva-ram those are details of our implementation, and not needed. But, here they are.
> 
> First, `TokenResponse`:
> 
> ```python
> from pydantic import BaseModel
> 
> class TokenResponse(BaseModel):
>     """An OAuth2 Token response."""
> 
>     access_token: str
>     token_type: str
>     expires_in: int
> ```
> 
> I believe the purpose was to improve the API generated documentation. However, it seems better to exclude the token-requesting endpoint from the API docs. See the discussion above for a few ways to do that.
> 
> `token_scheme` is:
> 
> ```python
> from fastapi.security import HTTPBasic
> token_scheme = HTTPBasic(auto_error=False)
> ```
> 
> This is the same ideas as [Simple HTTP Basic Auth](https://fastapi.tiangolo.com/advanced/security/http-basic-auth), but that code calls it `security` instead of `token_scheme`. I liked the more precise name, and wanted to allow something else to be called `security`.
> 
> `_token_settings` allows us to override some settings:
> 
> ```python
> def _token_settings(
>     settings: config.Settings = Depends(get_settings),
> ) -> Dict[str, Union[str, timedelta]]:
>     return {
>         "expires_delta": settings.token_expiration,
>         "secret_key": settings.secret_key,
>     }
> ```
> 
> `get_settings()` is basically the same as the one described in [Settings in a dependency](https://fastapi.tiangolo.com/advanced/settings/#settings-in-a-dependency). So, when run as an app, it gets these from the environment or defaults, and I can override just these two settings for tests.
> 
> Our defaults:
> 
> ```python
> from pydantic import BaseSettings
> 
> class Settings(BaseSettings):
>     secret_key: str
>     token_expiration: timedelta = timedelta(minutes=60)
>     ...
> ```
> 
> This means that `SECRET_KEY` must be set or config will fail. We use a long random string for each deployment. This concept was borrowed from Django, see [SECRET_KEY](https://docs.djangoproject.com/en/4.0/ref/settings/#secret-key) for how it is used there.

Isn't encoding secret_key in access_token insecure ? Anybody could decode with verify=False to retrieve it.

**Comment by jwhitlock** [comment]
> Isn't encoding secret_key in access_token insecure ? Anybody could decode with verify=False to retrieve it.

Yes, and we do not encode the `secret_key` in the access token. It is used 

@bdelpey the `secret_key` is used in `create_access_token`. Here's the version from [the FastAPI documentation](https://fastapi.tiangolo.com/tutorial/security/oauth2-jwt/?h=create_access_token#hash-and-verify-the-passwords):

```python
# to get a string like this run:
# openssl rand -hex 32
SECRET_KEY = "09d25e094faa6ca2556c818166b7a9563b93f7099f6f0f4caa6cf63b88e8d3e7"
ALGORITHM = "HS256"

def create_access_token(
    data: dict,
    expires_delta: Union[timedelta, None] = None
):
    to_encode = data.copy()
    if expires_delta:
        expire = datetime.utcnow() + expires_delta
    else:
        expire = datetime.utcnow() + timedelta(minutes=15)
    to_encode.update({"exp": expire})
    encoded_jwt = jwt.encode(to_encode, SECRET_KEY, algorithm=ALGORITHM)
    return encoded_jwt
```

See [jwt.encode](https://github.com/mpdavis/python-jose/blob/96474ecfb6ad3ce16f41b0814ab5126d58725e2a/jose/jwt.py#L13-L53) and [jws.py](https://github.com/mpdavis/python-jose/blob/master/jose/jws.py) for how `key=SECRET_KEY` is used (the `jose` docs are underwhelming).

Our code is open-source, so we didn't want to put the `SECRET_KEY` in the source code (and I think it is a bad idea for closed source code as well!), so we moved it to an environment variable and passed it into `create_access_token`. We also pass in an optional `now` override for testing purposes:

```python
def create_access_token(
    data: dict,
    expires_delta: timedelta,
    secret_key: str,
    now: Optional[datetime] = None,
) -> str:
    """Create a JWT string to act as an OAuth2 access token."""
    to_encode = data.copy()
    expire = (now or datetime.utcnow()) + expires_delta
    to_encode["exp"] = expire
    encoded_jwt = jwt.encode(to_encode, secret_key, algorithm="HS256")
    return encoded_jwt
```

If you want the whole code in context, see https://github.com/mozilla-it/ctms-api/pull/87.

**Comment by bvandewe** [comment]
Thanks a bunch everyone! This helped a lot to write a sample app that supports `Resource Owner Password Credentials` OR `Client Credentials` Grant types.

It uses [Keycloak](https://www.keycloak.org/) as the Authorization Server (which is great to keep the API responsibilities focused on validating tokens - never creating them).

https://github.com/bvandewe/securedapi

Hope it helps someone!

**Comment by Vortexdude** [comment]
Hello @MarlieChiller,
How I can create a flow that will take the `email` instead of `username` that comes with `OAuth2PasswordRequestForm`.

``` python
class Oauth2ClientCredentials(OAuth2):
    def __init__(
        self,
        tokenUrl: str,
        scheme_name: str = None,
        scopes: dict = None,
        auto_error: bool = True,
    ):
        if not scopes:
            scopes = {}
        flows = OAuthFlowsModel(password={"tokenUrl": tokenUrl, "scopes": scopes})
        super().__init__(flows=flows, scheme_name=scheme_name, auto_error=auto_error)

    async def __call__(self, request: Request) -> Optional[str]:
        authorization: str = request.headers.get("Authorization")
        scheme, param = get_authorization_scheme_param(authorization)
        if not authorization or scheme.lower() != "bearer":
            if self.auto_error:
                raise HTTPException(
                    status_code=HTTP_401_UNAUTHORIZED,
                    detail="Not authenticated",
                    headers={"WWW-Authenticate": "Bearer"},
                )
            else:
                return None
        return param

```

**Comment by AlexzPurewoko** [comment]
Here's the updated version to be used with OAuth2 FastAPI (from my implementation)

```python

from fastapi.openapi.models import OAuthFlows, OAuthFlowClientCredentials
from fastapi.security import OAuth2
from fastapi import HTTPException, Request, Form
from fastapi.security.utils import get_authorization_scheme_param
from typing import Optional
from starlette.status import HTTP_401_UNAUTHORIZED


class Oauth2ClientCredentials(OAuth2):
    def __init__(
        self,
        token_url: str,
        scheme_name: str = None,
        scopes: dict = None,
        auto_error: bool = True,
    ):
        if not scopes:
            scopes = {}
        flows = OAuthFlows(clientCredentials=OAuthFlowClientCredentials(tokenUrl=token_url, scopes=scopes))
        super().__init__(flows=flows, scheme_name=scheme_name, auto_error=auto_error)

    async def __call__(self, request: Request) -> Optional[str]:
        authorization: str = request.headers.get("Authorization")
        scheme, param = get_authorization_scheme_param(authorization)
        if not authorization or scheme.lower() != "bearer":
            if self.auto_error:
                raise HTTPException(
                    status_code=HTTP_401_UNAUTHORIZED,
                    detail="Not authenticated",
                    headers={"WWW-Authenticate": "Bearer"},
                )
            else:
                return None
        return param


class OAuth2ClientCredentialsRequestForm:
    def __init__(
        self,
        grant_type: str = Form(None, regex="client_credentials"),
        scope: str = Form(""),
        client_id: Optional[str] = Form(None),
        client_secret: Optional[str] = Form(None),
    ):
        self.grant_type = grant_type
        self.scopes = scope.split()
        self.client_id = client_id
        self.client_secret = client_secret
```
